### PR TITLE
Register custom scans

### PR DESCRIPTION
**Pull request by onderkalaci**
While working with `sql_monster` (#1804), we realized that Postgres complains about not knowing about Citus' custom scan nodes. It appears that we should explicitly register the nodes.

The PR includes some refactoring on the custom scan related codes given that they were scattered and it was a bit hard to follow the code.

### CODE REVIEW

File: src/backend/distributed/executor/multi_task_tracker_executor.c [hunk 1]
```diff
@@ -3004,3 +3007,37 @@ TrackerHashDisconnect(HTAB *taskTrackerHash)
...
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
+	{
+		DistributedPlan *distributedPlan = scanState->distributedPlan;
+		Job *workerJob = distributedPlan->workerJob;
+
+		/* we are taking locks on partitions of partitioned tables */
+		LockPartitionsInRelationList(distributedPlan->relationIdList, AccessShareLock);
```
Comment (marcocitus) [inline src/backend/distributed/executor/multi_task_tracker_executor.c]: I've been wondering, why isn't this in `CitusSelectBeginScan`?

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+	.BeginCustomScan = CitusSelectBeginScan,
+	.ExecCustomScan = CoordinatorInsertSelectExecScan,
+	.EndCustomScan = CitusEndScan,
+	.ReScanCustomScan = CitusReScan,
+	.ExplainCustomScan = CoordinatorInsertSelectExplainScan
+};
+
+
+/*
+ * Let PostgreSQL know about Citus' custom scan nodes.
```
Comment (marcocitus) [inline src/backend/distributed/executor/citus_custom_scan.c]: "RegisterCitusCustomScanMethods lets PostgreSQL know ..."

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+/*
+ * Load data collected by real-time or task-tracker executors into the tuplestore
+ * of CitusScanState. For that, we first create a tuple store, and then copy the
+ * files one-by-one into the tuple store.
+ *
+ * Note that in the long term it'd be a lot better if Multi*Execute() directly
+ * filled the tuplestores, but that's a fair bit of work.
+ */
+void
+LoadTuplesIntoTupleStore(CitusScanState *citusScanState, Job *workerJob)
```
Comment (marcocitus) [inline src/backend/distributed/executor/citus_custom_scan.c]: Does this function really belong here? It seems more of a generic executor utility function  (multi_server_executor?).

For example, we'll use part of this logic for the `read_records_file` function.

Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: Makes sense, moving to `multi_executor` because there is an interdependency among`multi_server_executor ` and `citus_custom_scan`

File: src/backend/distributed/executor/multi_task_tracker_executor.c (hunk 1 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/multi_task_tracker_executor.c]: I saw it in your PR, so didn't want to interfere with your changes. Also, since this commit only moves files rather than modifying, I'd like to keep it as it is.
